### PR TITLE
gr-uhd: add TwinRX LO sharing arguments to uhd_app

### PR DESCRIPTION
**Pull request by noc0lour**
Enables specification of --lo-export {None, True, False} and --lo-source {None, external, internal, companion} in uhd_app applications for usage with a TwinRX.
Specification of None skips specification of LO export/source features for a channel.

### CODE REVIEW

**Comment by jmcorgan** [comment]
Are there any version/ABI issues with respect to UHD for this?  Also, is this applicable unchanged between master (3.7.x) and next (3.8) branches in GR?

Finally, @mbr0wn, please review and approve.

**Comment by noc0lour** [comment]
UHD 3.10.0.0 is the minimum version for TwinRX support. Should be safe to use with UHD versions prior to that because nothing happens if you don't specify TwinRX specific commandline arguments. But it could confuse people. I'll see if I can apply these changes on next.

File: gr-uhd/apps/uhd_app.py [hunk 1]
```diff
@@ -204,6 +213,32 @@ def setup_usrp(self, ctor, args, cpu_format='fc32'):
...
         self.has_lo_sensor = 'lo_locked' in self.usrp.get_sensor_names()
+        # Set LO export and LO source operation
+        if (args.lo_export is not None) and (args.lo_source is not None):
+            (args.lo_source,args.lo_export) = self.normalize_lo_source_export_sel(args)
+            for chan,lo_source,lo_export in zip(self.channels,args.lo_source,args.lo_export):
+                if (lo_source == "None") or (lo_export == "None"):
+                    continue
+                if lo_export == "True":
+                    #If channel is LO source set frequency and store response
+                    self.usrp.set_lo_export_enabled(True, uhd.ALL_LOS, chan)
```
Comment (mbr0wn) [inline gr-uhd/apps/uhd_app.py]: Will this work on UHD 3.5?

File: gr-uhd/apps/uhd_app.py [hunk 2]
```diff
@@ -204,6 +213,32 @@ def setup_usrp(self, ctor, args, cpu_format='fc32'):
...
+                if lo_source == "internal":
+                    self.lo_source_channel = chan
+                    tune_resp = self.usrp.set_center_freq(treq,chan)
+                self.usrp.set_lo_source(lo_source, uhd.ALL_LOS,chan)
+            # Use lo source tune response to tune dsp_freq on remaining channels
+            if hasattr(args, 'lo_offset') and args.lo_offset is not None:
```
Comment (mbr0wn) [inline gr-uhd/apps/uhd_app.py]:
```{.py}
if getattr(args, 'lo_offset', None) is not None:
```

**Comment by mbr0wn** [comment]
This extends the Python API of `uhd_app`, and hence is not ABI-affecting. I think this should merge cleanly into next (@noc0lour can you plz confirm?).

I'm not so sure about being fine with older UHDs, because `ALL_LOS` was not defined there.

**Comment by mbr0wn** [comment]
...and neither is `set_lo_export_enabled()`.

**Comment by mbr0wn** [comment]
Also, the argument that people shouldn't call TwinRX-specific commands is obscured by the fact that `uhd_fft` (and other tools) will expose these new arguments. It's not obvious that your current setup doesn't support the LO API. Therefore, checking for existence of API calls, and adding some checks (or try/catches) would give a meaningful error rather than a 'attribute not defined' error.

**Comment by noc0lour** [comment]
Applies cleanly on master and next. 
If I try to execute uhd_fft with --lo-source and --lo-export specified I get this error:
```
return _uhd_swig.usrp_source_sptr_set_lo_export_enabled(self, enabled, name, chan)
RuntimeError: not implemented in this version
```
Should be verbose enough to leave it there or will people confuse it with gnuradio version?

**Comment by jmcorgan** [comment]
I'd prefer if the error message indicates UHD version needed to support that function call.

**Comment by noc0lour** [comment]
I'll catch the RuntimeError now and print following and exit:
```
[ERROR] LO sharing is not implemented in UHD versions prior 3.10.0.0. Please upgrade your UHD version.
```

**Comment by jmcorgan** [comment]
@mbr0wn are we good on this now?

**Comment by mbr0wn** [comment]
One tiny comment left :speech_balloon:

**Comment by noc0lour** [comment]
replaced `hasattr` with `getattr` as suggested.

**Comment by mbr0wn** [comment]
:+1:

**Pull request merged**